145 string to value conversions misinterpret number formats

mstranne opened the pull request:
### SUMMARY
Implements necessary changes for #145 

*Reviewers:* @numere-org/maintainers

### IMPLEMENTATION
* Implementation: 
Extension of the `isConvertible()` function in the check for the `ValueType`. 
The new struct `NumberFormatsVoter` is used to detect Number Formats in a string or further detect a common Format of in number of strings. This allows for a clear classification of the number format in `detectCommonType(...)`, ensuring correct conversion.

The implementation identifies thousand and decimal separators. The new functionality, `strChangeNumberFormat()`, is then used in the `Convert` function of the `TableColumn` class for proper conversion.

* Implementation test: (*Describe the type of test, which you performed, and if it was successful*)
IN SW-Tests see TableTests. 
There are four ways in which separator characters can be unambiguously associated with a number format. I have written tests for all possible number formats for each of these types. This includes EU and US styles with both spaces and special characters as thousand separators. All the tests are working correctly.

### DOCUMENTATION
* [x] ChangesLog updated
* [x] Code changes commented
* **Documentation articles:**
    * [ ] corresponding documentation articles updated
    * [ ] new documentation articles created
    * [x] not needed
* **Language files:**
    * [ ] corresponding language files updated
    * [x] not needed

----------------

### TESTS BY REVIEWERS
* [x] Added to the automatic SW tests
* [ ] Tested manually

numeredev commented on kernel/core/utils/stringtools.cpp:
```diff
@@ -1016,18 +1050,129 @@ bool isConvertible(const std::string& sStr, ConvertibleType type)
+                    inNum = true;
+                    numStartIdx = i;
+                }
+                if(sStr[i] == ',' || sStr[i] == '.' || sStr[i] == ' ' ) {
```
This huge block appears to replicate a lookup table. If you could transform it in such a thing, that would greatly improve readability (and testability). I think, you're familiar with the concept of cyclomatic complexity. I counted a 17 for this block, which is beyond acceptable limits.

Consider the following possible solutions:

* Create a lookup table, where you can search for thousand seps and decimal seps and the table will return you the corresponding bitflags
* Create a map of all possible combinations (linear lookup table), if that is handleable, and use `std::map::find()` to look for the correct number format, e.g., `auto iter = numberFormats.find("xxx,xxx.xxx"); if (iter != numberFormats.end()) return iter->second`
* Create a simple structure, where you can set the found separator in their order and the structure will then solve the number format for you

numeredev commented on kernel/core/utils/stringtools.cpp:
```diff
@@ -24,6 +24,7 @@
 #include <cstring>
 #include <sstream>
 #include <iomanip>
+#include <regex>
```
Is not needed anymore, is it?

numeredev commented on kernel/core/utils/stringtools.cpp:
```diff
@@ -999,8 +1034,7 @@ static bool isDateTimePattern(const std::string& sStr, size_t pos)
 /////////////////////////////////////////////////
-bool isConvertible(const std::string& sStr, ConvertibleType type)
-{
+bool isConvertible(const std::string& sStr, ConvertibleType type){
```
Consider the following adaption of this function:

```suggestion
bool isConvertible(const std::string& sStr, ConvertibleType type, NumberFormatsVoter* voter = nullptr){
```

Here, `NumberFormatsVoter` is a structure/class to be designed replacing your two global variables. It may also contain the `vote()` method from further up and will only be used if non-zero.

```C++
struct NumberFormatsVoter
{ // Preferring ANSI brace format ;-)
    // necessary attributes here

    void vote(int numType)
    {
        // ...
    }

    int getFormat() const
    {
        // ...
    }
}
```

numeredev commented on kernel/core/utils/stringtools.cpp:
```diff
@@ -1016,18 +1050,129 @@ bool isConvertible(const std::string& sStr, ConvertibleType type)
+        }
+
+        //TODO Check last value for number format
+        voteNumType(nFormat);
```
Like so
```suggestion
        if (voter)
            voter->vote(nFormat);
```

numeredev commented on kernel/core/utils/stringtools.cpp:
```diff
@@ -989,6 +990,40 @@ static bool isDateTimePattern(const std::string& sStr, size_t pos)
+static void voteNumType(int numType)  {
+    if((numType & NUM_DECIMAL_EU && numType & NUM_K_US) ||
+       (numType & NUM_DECIMAL_US && numType & NUM_K_EU)) {
+        num_format_votes[5]++;  // INVALID Combination
```
You do not want to use number literals here. There might be furthe formats in the future. Consider a solution like so:

```C++
enum
{
    NUM_DECIMAL_EU,
    // all other format types
    NUM_FORMAT_COUNT // This will be the number of formats available (and therefore scale with further formats)
}

// the declaration of the necessary array like so (will scale automatically)
int num_format_votes[NUM_FORMAT_COUNT];

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -213,6 +227,12 @@ std::string ellipsize(const std::string& sLongString, size_t nMaxStringLength =
+// NEW
+extern int last_num_format;
+extern int num_format_votes[];
+static void voteNumType(int numType);
```
You do not declare `static` functions in a header (with rare exceptions*). The purpose of `static` functions is to make them local-only. This has the benefit of having lower interdependencies as well as the compiler is able to optimize better (bc. it knows all possible usages of this function).

(* There are cases, when `static` functions are written in a header file, that is: this header file is an actual implementation file and only used in a single other implementation file.)

numeredev commented on kernel/core/datamanagement/tablecolumnimpl.cpp:
```diff
@@ -34,6 +36,10 @@ static ConvertibleType detectCommonType(const std::vector<std::string>& vVals)
 {
     ConvertibleType convType = CONVTYPE_NONE;
 
+    last_num_format = 0;
+    for(int i = 0; i < 6; i++)
+        num_format_votes[i] = 0;
```
This will simplify to

```suggestion
    int last_num_format = 0; // is actually only used here
    NumberFormatsVoter voter;
```

numeredev commented on kernel/core/datamanagement/tablecolumnimpl.cpp:
```diff
@@ -77,10 +83,39 @@ static ConvertibleType detectCommonType(const std::vector<std::string>& vVals)
+        last_num_format = 0;
+
+        // NEW check voting
+        if(num_format_votes[5] > 0) {  // if one invalid, all invalid ?
+            last_num_format = NUM_INVALID;
+            return CONVTYPE_NONE;
+        }
+
+        if((num_format_votes[2] > num_format_votes[3]) && (num_format_votes[2] > num_format_votes[4]))
+            last_num_format |= NUM_K_EU;
+        else if((num_format_votes[3] > num_format_votes[2]) && (num_format_votes[3] > num_format_votes[4]))
+            last_num_format |= NUM_K_US;
+        else if(num_format_votes[4] > 0)
+            last_num_format |= NUM_K_SPACE;
+
+        int add = 0;
+        if(last_num_format == 0)
+            add = 1;
+
+        if(num_format_votes[0] > num_format_votes[1])
+            last_num_format |= NUM_DECIMAL_EU | (NUM_K_EU * add);
+        else if(num_format_votes[1] > 0)
+            last_num_format |= NUM_DECIMAL_US | (NUM_K_US * add);
```
Here as well: that will be shortened to something like

```C++
last_num_format = voter.getFormat();
// ...
```

numeredev reviewed (changes requested):
I've added a bunch of suggestions for you to consider. Get in touch, if something is unclear.

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -130,6 +131,151 @@ enum TimeDateFormat
+        int num_format = 0;
+
+        // NEW check voting
+        if(num_format_votes[5] > 0) {  // if one invalid, all invalid ?
```
See above about numeric literals

numeredev reviewed (changes requested):
I've some more feedback, but it starts to get somewhere 😉

mstranne commented on kernel/core/utils/stringtools.cpp:
```diff
@@ -1016,18 +1018,41 @@ bool isConvertible(const std::string& sStr, ConvertibleType type)
-        // Regression fix introduced because NA is accepted as NaN
-        for (size_t i = 1; i < sStr.length()-1; i++)
+        // NEW
+        bool inNum = false;
```
How do you mean?

The string must be iterated anyway for this check:

```
  if (i > 0 && i-1 < sStr.length() && (sStr[i] == '-' || sStr[i] == '+'))
                {
                    if (tolower(sStr[i-1]) != 'e'
                        && (isdigit(sStr[i-1]) && sStr.find_first_of("iI", i+1) == std::string::npos)
                        && sStr[i-1] != ' ')
                        return false;
                }
```

Only the first and last char can be skipped if voter is NULL.

numeredev replied:
Right, have overseen that the regression fix was moved inside the new logics. Then my proposal might not be fitting.

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -130,6 +131,241 @@ enum TimeDateFormat
+        m_last_idx = idx;
+
+        // max 3 symbols at ones in tape
+        while(m_tape.length() > 3)
+            m_tape.erase(m_tape.begin());
```
What about this?
```suggestion
        if (m_tape.length() > 3)
            m_tape.erase(0, m_tape.length()-3);
```
Should be faster as only one reallocation is necessary.

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -130,6 +131,241 @@ enum TimeDateFormat
+    /// \return int
+    ///
+    /////////////////////////////////////////////////
+    int endParseNumber(int idx)
```
This function could automatically call `vote()`, I think

mstranne replied:
Yes!
I did not call it because someone might only be interested in finding the number format of a single string, which would not require the voting functionality at all.

numeredev replied:
Yeah, valid reason. Keep it as it is, or provide a wrapper like `void endParseAndVote(int idx)` or `void finalize(int idx)` or so.

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -130,6 +131,241 @@ enum TimeDateFormat
+    /////////////////////////////////////////////////
+    void vote(int numType)
+    {
+        if((numType & NUM_DECIMAL_EU && numType & NUM_K_US) ||
+                (numType & NUM_DECIMAL_US && numType & NUM_K_EU))
```
```suggestion
        if (numType & (NUM_DECIMAL_EU | NUM_K_US)
            || numType & (NUM_DECIMAL_US | NUM_K_EU))
```
Might be easier to read

numeredev replied:
Forget that, that's not the same thing as the original code

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -130,6 +131,241 @@ enum TimeDateFormat
+    int getFormat()
+    {
+        // As soon as there are more than 2 set, we will have at least one bad sample that would be parsed wrong
+        int inval_cnt = 0;
+        for(int i = 0; i < FMT_COUNT; i++)
+        {
+            if(m_num_format_votes[i] > 0)
+                inval_cnt++;
+        }
+        if(inval_cnt > 2 || m_num_format_votes[FMT_INVALID] > 0)
+            return NUM_INVALID;
```
Looks a bit over-engineered. Why do we need to count for number of votes? Isn't that enough:
```suggestion
        if (m_num_format_votes[FMT_INVALID])
            return NUM_INVALID;
```

mstranne replied:
If our objective is to obtain the format with the highest number of votes, then yes. 
However, if there are votes for both the US and the EU, it is certain that one of them will be parsed incorrectly. The idea behind the "> 2" check was to not convert anything when different formats emerge within the same column.

numeredev replied:
I'm not sure, whether I understood that correctly, but how does adding up *all* votes help with the idea to filter out invalid combinations? And then: if only one format is invalid, then the result of the whole vote is invalid, so there's no reason in *counting* the votes (at this point).

mstranne replied:
It is not adding up all votes, it is a counter of how many different votes we have.
Then the idea was that we should max have 2 different types of votes, decimal & thousands separator.
However, as this would consider two decimal or two thousand separators as valid, this code does not work as we want.

Just for clarification, if we encounter more than one number format in an entire column, we don't want to proceed with the conversion to `CONVTYPE_VALUE` at all, correct? In that case, we should return `NUM_INVALID` here.

numeredev replied:
You're right, I was probably confused by the variable name. If those two number formats exclude each other and there's no possibility of a false detection, yes.

mstranne replied:
Oh true thats a good point.
When only decimal separators occur, it is not a problem to parse both correctly!

I will change that.

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -130,6 +131,241 @@ enum TimeDateFormat
+        else if(m_num_format_votes[FMT_DEC_US] > 0)
+            num_format |= NUM_DECIMAL_US | (NUM_K_US * add);
+
+        return num_format;
```
You always assume, there's a strong order within your votes, which might not be true: `1 000 000,00; 1,000,000.00`. Each format now appears the same time, so it might happen that `num_format` is still 0 at this point. What does 0 mean? Isn't that just the same thing as `NUM_INVALID`?

mstranne replied:
Good Point!
I did assume that we want to return NUM_INVALID when two Formats accure in the > 2 check before.

`num_format ` = 0 would mean no unique Number Format found, but also no INVALID combination did appear. 
This means, the convertion Function will do nothing. This would happen if all nums are without any seperators at all, or all are 10.000  or 1,000 where it could not be specified if thousans or decimal seperator.

numeredev replied:
OK, but isn't that in the end the same thing (for conversion) as `NUM_INVALID`. Semantically speaking, no format is also an invalid format.

Regarding the last point: if you cannot decide, whether EU or US, the default should be the US format for numbers.

mstranne replied:
Returning `NUM_INVALID` should set `CONVTYPE_NONE`, then no conversion is done. 
When 0 is returned `strChangeNumberFormat(...)`, would currently do nothing and `StrToCmplx(...)` is called with the string as it is. 

But I guess we should not return 0, but US format right? 
We should only return 0 in two cases:
- No separators at all, so no conversation needed.
- We only find separators ',' or '.' with no more than 3 digits left and exactly 3 digits right. (eg. 100.000 or 10,000)

numeredev replied:
Following my previous comment, I'd appreciate if it defaults to US format but allows for EU format together, if none of them are using their thousands separator. If that is the same thing as return 0 in your implementation (with the addition that `replaceall(STR, ",", ".")` is called), then that's the way to go.

numeredev reviewed (changes requested):
I've only some minor things to consider and did also find a probable bug within your logic. Didn't test that yet, but I think that this is the last iteration.

mstranne commented:
I changed the `getFormat()` such that the counter now is incremented each time it sets one format instead of the loop I had before.

Also I noticed a bug! 

`num_format |= NUM_DECIMAL_US | (NUM_K_US * add);`

here I assumed that if only DECIMAL is set, we can also set the corresponding K separator. But here SPACE and US are both valid thousand separators. 

Now I added some further checks for the ambiguous cases when we have 1 separator and exactly 3 digits afterwards. 
Since SPACE is only allowed as thousands separator, `1 000` is actually a unique pattern and can be used.

mstranne commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -343,34 +354,48 @@ struct NumberFormatsVoter
+        // IF we have some ambigious votes that correspont to unique decimals use these to set thousands seperator.
+        if(m_num_format_votes[FMT_DEC_EU] > 0)
+        {
+            num_format |= NUM_DECIMAL_EU | (m_num_format_votes[FMT_AMBIGIOUS_DOT] > 0 ? NUM_K_EU : 0);
```
vote for these ambiguous cases and if at least one unique DECIMAL separator was found and also ambiguous thousands separators where found, set also the thousand separator.

numeredev replied:
That depends on how easy that would be implemented. Consider that a fancy solution might be hard to understand and therefore hard to maintain. If the ambiguous votes do not really impact performance or change correctness, I'd vote for keeping them.

numeredev commented on kernel/core/utils/stringtools.hpp:
```diff
@@ -130,6 +131,275 @@ enum TimeDateFormat
+    /////////////////////////////////////////////////
+    void startParseNumber(int idx)
+    {
+        m_last_idx = idx-1; //todo wie handeln wir 1ste und letzte, da diese nicht seperatoren sind
```
Is this todo still valid?

numeredev reviewed (approved):
Changes accepted

numeredev commented:
@mstranne  Found some bugs while adding "negative" tests. Have a look into them.

numeredev reviewed (changes requested):
There's a bug in the negative cases (no conversion shall happen).
